File: backend/app/api/admin/node_configs.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from typing import List, Optional

from app.core.database import get_db
from app.core.deps import get_current_admin_user
from app.models.user import User
from app.models.node_config import NodeConfig
from app.models.llm_config import LLMConfig
from app.models.prompt_template import PromptTemplate
from app.models.fastclaw_agent_config import FastClawAgentConfig
from app.models.skill_agent_config import SkillAgentConfig
from app.modules.bookplate.node_types import NODE_TEMPLATE_MAP
from app.schemas.admin import (
    NodeConfigCreate,
    NodeConfigUpdate,
    NodeConfigOut,
    NodeConfigReorderPayload,
)

router = APIRouter(prefix="/admin/node-configs", tags=["admin-node-configs"])


def _validate_node_type(node_type: str) -> None:
    """校验节点模板类型是否为代码内置模板。"""
    if node_type not in NODE_TEMPLATE_MAP:
        raise HTTPException(status_code=400, detail=f"未知的节点模板类型: {node_type}")


def _validate_mode(
    has_llm: bool,
    has_prompt: bool,
    has_agent: bool,
    has_skill_agent: bool,
) -> None:
    """校验模式互斥：agent / skill_agent 模式与「提示词 + 大模型」只能选择一组。"""
    if (has_agent or has_skill_agent) and (has_llm or has_prompt):
        raise HTTPException(
            status_code=400,
            detail="Agent / Skill Agent 模式与「提示词 + 大模型」互斥，只能选择一组",
        )
    if has_agent and has_skill_agent:
        raise HTTPException(
            status_code=400,
            detail="FastClaw Agent 与 Skill Agent 模式互斥，只能选择一组",
        )


def _validate_refs(
    db: Session,
    llm_config_id: Optional[int],
    prompt_id: Optional[int],
    agent_config_id: Optional[int],
    skill_agent_config_id: Optional[int],
) -> None:
    if llm_config_id is not None:
        if not db.query(LLMConfig).filter(LLMConfig.id == llm_config_id).first():
            raise HTTPException(status_code=400, detail="所选模型配置不存在")
    if prompt_id is not None:
        if not db.query(PromptTemplate).filter(PromptTemplate.id == prompt_id).first():
            raise HTTPException(status_code=400, detail="所选提示词模板不存在")
    if agent_config_id is not None:
        if not db.query(FastClawAgentConfig).filter(FastClawAgentConfig.id == agent_config_id).first():
            raise HTTPException(status_code=400, detail="所选 FastClaw Agent 配置不存在")
    if skill_agent_config_id is not None:
        if not db.query(SkillAgentConfig).filter(SkillAgentConfig.id == skill_agent_config_id).first():
            raise HTTPException(status_code=400, detail="所选 Skill Agent 配置不存在")


def _to_out(nc: NodeConfig) -> NodeConfigOut:
    return NodeConfigOut(
        id=nc.id,
        node_type=nc.node_type,
        name=nc.name,
        group=nc.group,
        group_order=nc.group_order,
        llm_config_id=nc.llm_config_id,
        prompt_id=nc.prompt_id,
        agent_config_id=nc.agent_config_id,
        skill_agent_config_id=nc.skill_agent_config_id,
        llm_config_name=nc.llm_config.name if nc.llm_config else None,
        prompt_name=nc.prompt.name if nc.prompt else None,
        agent_config_name=nc.agent_config.name if nc.agent_config else None,
        skill_agent_config_name=nc.skill_agent_config.name if nc.skill_agent_config else None,
        agent_config_agent_name=nc.agent_config.agent_name if nc.agent_config else None,
        is_active=nc.is_active,
        created_at=nc.created_at,
        updated_at=nc.updated_at,
    )


def _load_nc(db: Session, nc_id: int) -> NodeConfig:
    nc = db.query(NodeConfig).filter(NodeConfig.id == nc_id).first()
    if not nc:
        raise HTTPException(status_code=404, detail="节点配置不存在")
    return nc


@router.get("", response_model=List[NodeConfigOut])
def list_node_configs(
    node_type: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """节点配置列表（可按 node_type 过滤）。"""
    query = db.query(NodeConfig)
    if node_type:
        query = query.filter(NodeConfig.node_type == node_type)
    items = query.order_by(NodeConfig.id.asc()).all()
    return [_to_out(nc) for nc in items]


@router.post("", response_model=NodeConfigOut)
def create_node_config(
    payload: NodeConfigCreate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """创建节点配置：同一节点模板可创建多条配置（= 画板中的多个节点变体）。"""
    _validate_node_type(payload.node_type)
    if not payload.name.strip():
        raise HTTPException(status_code=400, detail="节点名称不能为空")
    _validate_mode(
        payload.llm_config_id is not None,
        payload.prompt_id is not None,
        payload.agent_config_id is not None,
        payload.skill_agent_config_id is not None,
    )
    _validate_refs(
        db,
        payload.llm_config_id,
        payload.prompt_id,
        payload.agent_config_id,
        payload.skill_agent_config_id,
    )
    nc = NodeConfig(**payload.model_dump())
    db.add(nc)
    db.commit()
    db.refresh(nc)
    db.expire(nc)
    return _to_out(nc)


@router.patch("/{nc_id}", response_model=NodeConfigOut)
def update_node_config(
    nc_id: int,
    payload: NodeConfigUpdate,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """修改节点配置（合并现有字段后校验模式互斥与引用有效性）。"""
    nc = _load_nc(db, nc_id)
    data = payload.model_dump(exclude_unset=True)
    if "node_type" in data and data["node_type"] is not None:
        _validate_node_type(data["node_type"])
    if "name" in data and data["name"] is not None and not str(data["name"]).strip():
        raise HTTPException(status_code=400, detail="节点名称不能为空")

    merged = {
        "node_type": nc.node_type,
        "llm_config_id": nc.llm_config_id,
        "prompt_id": nc.prompt_id,
        "agent_config_id": nc.agent_config_id,
        "skill_agent_config_id": nc.skill_agent_config_id,
        **data,
    }
    _validate_mode(
        merged.get("llm_config_id") is not None,
        merged.get("prompt_id") is not None,
        merged.get("agent_config_id") is not None,
        merged.get("skill_agent_config_id") is not None,
    )
    _validate_refs(
        db,
        merged.get("llm_config_id"),
        merged.get("prompt_id"),
        merged.get("agent_config_id"),
        merged.get("skill_agent_config_id"),
    )
    for key, value in data.items():
        setattr(nc, key, value)
    db.add(nc)
    db.commit()
    db.refresh(nc)
    db.expire(nc)
    return _to_out(nc)


@router.post("/reorder-groups")
def reorder_node_groups(
    payload: NodeConfigReorderPayload,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """批量更新自定义分组排序：按 group 名写入排序序号（同组所有配置共享）。

    接收完整分组顺序列表（仅含自定义分组）；未出现的分组序号不变。
    """
    seen: List[str] = []
    for item in payload.groups:
        g = item.group.strip()
        if not g or g in seen:
            raise HTTPException(status_code=400, detail="分组名不能为空或重复")
        seen.append(g)
        for nc in db.query(NodeConfig).filter(NodeConfig.group == g).all():
            nc.group_order = item.order
    db.commit()
    return {"message": "分组排序已更新"}


@router.delete("/{nc_id}")
def delete_node_config(
    nc_id: int,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_admin_user),
):
    """删除节点配置（画板「+」菜单中将不再出现该节点变体）。"""
    nc = _load_nc(db, nc_id)
    db.delete(nc)
    db.commit()
    return {"message": "节点配置已删除"}

File: backend-ts/src/api/admin/node-configs.ts
import { eq } from 'drizzle-orm';
import type { FastifyInstance } from 'fastify';
import { getDb, type DB } from '../../config/database.js';
import {
  fastclawAgentConfigs,
  llmConfigs,
  nodeConfigs,
  promptTemplates,
  skillAgentConfigs,
} from '../../db/schema.js';
import { NODE_TEMPLATES } from '../../modules/bookplate/node-types.js';
import { now, toIso } from '../../shared/datetime.js';

/**
 * 节点配置管理（对应 Python `app/api/admin/node_configs.py`）：
 * - GET/POST /api/admin/node-configs（列表/新建，同一节点模板可建多条配置）
 * - PATCH/DELETE /api/admin/node-configs/:id（修改/删除）
 * - POST /api/admin/node-configs/reorder-groups（批量更新自定义分组排序）
 *
 * 校验：节点模板类型必须为代码内置；模式互斥（Agent/Skill Agent 与「提示词+大模型」二选一）；
 * 引用字段必须指向已存在的配置。
 */

type NodeConfigRow = typeof nodeConfigs.$inferSelect;

interface NodeConfigPayload {
  node_type?: string | null;
  name?: string | null;
  group?: string | null;
  llm_config_id?: number | null;
  prompt_id?: number | null;
  agent_config_id?: number | null;
  skill_agent_config_id?: number | null;
  is_active?: boolean | null;
}

interface NodeConfigOut {
  id: number;
  node_type: string;
  name: string;
  group: string | null;
  group_order: number;
  llm_config_id: number | null;
  prompt_id: number | null;
  agent_config_id: number | null;
  skill_agent_config_id: number | null;
  llm_config_name: string | null;
  prompt_name: string | null;
  agent_config_name: string | null;
  skill_agent_config_name: string | null;
  agent_config_agent_name: string | null;
  is_active: boolean;
  created_at: string | null;
  updated_at: string | null;
}

function toOut(db: DB, nc: NodeConfigRow): NodeConfigOut {
  const llm = nc.llmConfigId != null ? db.select().from(llmConfigs).where(eq(llmConfigs.id, nc.llmConfigId)).get() : undefined;
  const prompt = nc.promptId != null ? db.select().from(promptTemplates).where(eq(promptTemplates.id, nc.promptId)).get() : undefined;
  const agent = nc.agentConfigId != null ? db.select().from(fastclawAgentConfigs).where(eq(fastclawAgentConfigs.id, nc.agentConfigId)).get() : undefined;
  const skillAgent = nc.skillAgentConfigId != null ? db.select().from(skillAgentConfigs).where(eq(skillAgentConfigs.id, nc.skillAgentConfigId)).get() : undefined;
  return {
    id: nc.id,
    node_type: nc.nodeType,
    name: nc.name,
    group: nc.group,
    group_order: nc.groupOrder,
    llm_config_id: nc.llmConfigId,
    prompt_id: nc.promptId,
    agent_config_id: nc.agentConfigId,
    skill_agent_config_id: nc.skillAgentConfigId,
    llm_config_name: llm?.name ?? null,
    prompt_name: prompt?.name ?? null,
    agent_config_name: agent?.name ?? null,
    skill_agent_config_name: skillAgent?.name ?? null,
    agent_config_agent_name: agent?.agentName ?? null,
    is_active: !!nc.isActive,
    created_at: toIso(nc.createdAt),
    updated_at: toIso(nc.updatedAt),
  };
}

function loadNodeConfig(db: DB, id: number): NodeConfigRow | undefined {
  return db.select().from(nodeConfigs).where(eq(nodeConfigs.id, id)).get();
}

function validateNodeType(nodeType: string): string | null {
  if (!NODE_TEMPLATES.some((t) => t.type === nodeType)) {
    return `未知的节点模板类型: ${nodeType}`;
  }
  return null;
}

/** 模式互斥校验：Agent / Skill Agent 与「提示词 + 大模型」只能选一组；两种 Agent 互斥。 */
function validateMode(hasLlm: boolean, hasPrompt: boolean, hasAgent: boolean, hasSkillAgent: boolean): string | null {
  if ((hasAgent || hasSkillAgent) && (hasLlm || hasPrompt)) {
    return 'Agent / Skill Agent 模式与「提示词 + 大模型」互斥，只能选择一组';
  }
  if (hasAgent && hasSkillAgent) {
    return 'FastClaw Agent 与 Skill Agent 模式互斥，只能选择一组';
  }
  return null;
}

function validateRefs(
  db: DB,
  llmConfigId: number | null | undefined,
  promptId: number | null | undefined,
  agentConfigId: number | null | undefined,
  skillAgentConfigId: number | null | undefined
): string | null {
  if (llmConfigId != null && !db.select().from(llmConfigs).where(eq(llmConfigs.id, llmConfigId)).get()) {
    return '所选模型配置不存在';
  }
  if (promptId != null && !db.select().from(promptTemplates).where(eq(promptTemplates.id, promptId)).get()) {
    return '所选提示词模板不存在';
  }
  if (agentConfigId != null && !db.select().from(fastclawAgentConfigs).where(eq(fastclawAgentConfigs.id, agentConfigId)).get()) {
    return '所选 FastClaw Agent 配置不存在';
  }
  if (skillAgentConfigId != null && !db.select().from(skillAgentConfigs).where(eq(skillAgentConfigs.id, skillAgentConfigId)).get()) {
    return '所选 DeepSeek Agent 配置不存在';
  }
  return null;
}

export async function registerNodeConfigsAdminRouter(app: FastifyInstance): Promise<void> {
  const admin = { preHandler: app.requireAdmin };

  // 列表（可按 node_type 过滤）
  app.get('/api/admin/node-configs', admin, async (request) => {
    const q = (request.query ?? {}) as { node_type?: string };
    const db = getDb();
    const query = q.node_type
      ? db.select().from(nodeConfigs).where(eq(nodeConfigs.nodeType, q.node_type))
      : db.select().from(nodeConfigs);
    return query.orderBy(nodeConfigs.id).all().map((nc) => toOut(db, nc));
  });

  // 新建
  app.post(
    '/api/admin/node-configs',
    admin,
    async (request, reply) => {
      const p = (request.body ?? {}) as NodeConfigPayload;
      const nodeType = p.node_type ?? '';
      if (nodeType) {
        const err = validateNodeType(nodeType);
        if (err) return reply.code(400).send({ detail: err });
      }
      if (!p.name?.trim()) return reply.code(400).send({ detail: '节点名称不能为空' });

      const modeErr = validateMode(
        p.llm_config_id != null,
        p.prompt_id != null,
        p.agent_config_id != null,
        p.skill_agent_config_id != null
      );
      if (modeErr) return reply.code(400).send({ detail: modeErr });
      const refErr = validateRefs(getDb(), p.llm_config_id, p.prompt_id, p.agent_config_id, p.skill_agent_config_id);
      if (refErr) return reply.code(400).send({ detail: refErr });

      const db = getDb();
      const row = db
        .insert(nodeConfigs)
        .values({
          nodeType,
          name: p.name.trim(),
          group: p.group ?? null,
          llmConfigId: p.llm_config_id ?? null,
          promptId: p.prompt_id ?? null,
          agentConfigId: p.agent_config_id ?? null,
          skillAgentConfigId: p.skill_agent_config_id ?? null,
          isActive: p.is_active ?? true,
          createdAt: now(),
          updatedAt: now(),
        })
        .returning()
        .get();
      return toOut(db, row);
    }
  );

  // 修改（合并现有字段后校验）
  app.patch(
    '/api/admin/node-configs/:id',
    admin,
    async (request, reply) => {
      const id = Number((request.params as { id: string }).id);
      const p = (request.body ?? {}) as NodeConfigPayload;
      const db = getDb();
      const nc = loadNodeConfig(db, id);
      if (!nc) return reply.code(404).send({ detail: '节点配置不存在' });

      if (p.node_type != null) {
        const err = validateNodeType(p.node_type);
        if (err) return reply.code(400).send({ detail: err });
      }
      if (p.name != null && !String(p.name).trim()) return reply.code(400).send({ detail: '节点名称不能为空' });

      const merged: NodeConfigPayload = {
        node_type: p.node_type ?? nc.nodeType,
        llm_config_id: p.llm_config_id !== undefined ? p.llm_config_id : nc.llmConfigId,
        prompt_id: p.prompt_id !== undefined ? p.prompt_id : nc.promptId,
        agent_config_id: p.agent_config_id !== undefined ? p.agent_config_id : nc.agentConfigId,
        skill_agent_config_id: p.skill_agent_config_id !== undefined ? p.skill_agent_config_id : nc.skillAgentConfigId,
      };
      const modeErr = validateMode(
        merged.llm_config_id != null,
        merged.prompt_id != null,
        merged.agent_config_id != null,
        merged.skill_agent_config_id != null
      );
      if (modeErr) return reply.code(400).send({ detail: modeErr });
      const refErr = validateRefs(db, merged.llm_config_id, merged.prompt_id, merged.agent_config_id, merged.skill_agent_config_id);
      if (refErr) return reply.code(400).send({ detail: refErr });

      const set: Record<string, unknown> = {};
      if (p.node_type != null) set.nodeType = p.node_type;
      if (p.name != null) set.name = p.name;
      if (p.group !== undefined) set.group = p.group;
      if (p.llm_config_id !== undefined) set.llmConfigId = p.llm_config_id;
      if (p.prompt_id !== undefined) set.promptId = p.prompt_id;
      if (p.agent_config_id !== undefined) set.agentConfigId = p.agent_config_id;
      if (p.skill_agent_config_id !== undefined) set.skillAgentConfigId = p.skill_agent_config_id;
      if (p.is_active !== undefined) set.isActive = p.is_active;
      set.updatedAt = now();
      db.update(nodeConfigs).set(set).where(eq(nodeConfigs.id, id)).run();
      return toOut(db, loadNodeConfig(db, id)!);
    }
  );

  // 批量更新自定义分组排序
  app.post(
    '/api/admin/node-configs/reorder-groups',
    admin,
    async (request, reply) => {
      const groups = (request.body as { groups?: { group?: string; order?: number }[] } | undefined)?.groups ?? [];
      const db = getDb();
      const seen = new Set<string>();
      for (const item of groups) {
        const g = (item.group ?? '').trim();
        if (!g || seen.has(g)) return reply.code(400).send({ detail: '分组名不能为空或重复' });
        seen.add(g);
        const order = item.order ?? 0;
        const rows = db.select().from(nodeConfigs).where(eq(nodeConfigs.group, g)).all();
        for (const nc of rows) {
          db.update(nodeConfigs).set({ groupOrder: order, updatedAt: now() }).where(eq(nodeConfigs.id, nc.id)).run();
        }
      }
      return { message: '分组排序已更新' };
    }
  );

  // 删除
  app.delete('/api/admin/node-configs/:id', admin, async (request, reply) => {
    const id = Number((request.params as { id: string }).id);
    const db = getDb();
    if (!loadNodeConfig(db, id)) return reply.code(404).send({ detail: '节点配置不存在' });
    db.delete(nodeConfigs).where(eq(nodeConfigs.id, id)).run();
    return { message: '节点配置已删除' };
  });
}
